src/storage/store.js

```diff
@@ -0,0 +1,77 @@
+import uuid from "uuid/v4";
+import { Validator } from "jsonschema";
+
+const LOCAL_STORE_KEY = "___mozilla_duck";
+const STORE_STATE_CACHE_KEY = Symbol();
+const validator = new Validator();
+
+// Durable (via local-storage) schema-enforced state that is meant to be consumed via forward data flow.
+// (Think flux but with way less incidental complexity, at least for now :))
+export const SCHEMA = {
+  id: "/MozillaDuckStore",
+
+  definitions: {
+    profile: {
+      type: "object",
+      additionalProperties: false,
+      properties: {
+        display_name: { type: "string", pattern: "^[A-Za-z0-9-]{3,32}$" },
+      }
+    }
+  },
+
+  type: "object",
+
+  properties: {
+    id: { type: "string", pattern: "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$" },
+    profile: { "$ref": "#/definitions/profile" },
+  },
+
+  additionalProperties: false
+}
+
+export default class Store {
+  subscribers = new Set();
+
+  constructor() {
+    if (localStorage.getItem(LOCAL_STORE_KEY) === null) {
+      localStorage.setItem(LOCAL_STORE_KEY, JSON.stringify({ id: uuid() }));
+    }
+  }
+
+  get state() {
+    if (!this.hasOwnProperty(STORE_STATE_CACHE_KEY)) {
+      this[STORE_STATE_CACHE_KEY] = JSON.parse(localStorage.getItem(LOCAL_STORE_KEY));
+    }
+
+    return this[STORE_STATE_CACHE_KEY];
+  }
+
+  subscribe(subscriber) {
+    this.subscribers.add(subscriber);
+  }
```

Comment:
Couldn't we extend from EventTarget instead?

Reply:
yup didn't have that dependency at the time i wrote this